reark/src/main/java/io/reark/reark/data/stores/cores/ContentProviderStoreCoreBase.java

```diff
@@ -89,38 +145,85 @@ public static Handler createHandler(@NonNull final String name) {
+                Observable.amb(
+                        stream.debounce(groupingTimeout, TimeUnit.MILLISECONDS),
+                        stream.skip(groupMaxSize - 1))
+                        .first() // Complete observable after the first reached trigger
+                        .repeatWhen(observable -> observable))); // Resubscribe immediately for the next buffer
+    }
 
-        if (cursor != null) {
-            if (cursor.moveToFirst()) {
-                U currentItem = read(cursor);
-                valuesEqual = newItem.equals(currentItem);
-
-                if (!valuesEqual) {
-                    Log.v(TAG, "Merging values at " + pair.second);
-                    newItem = mergeValues(currentItem, newItem);
-                    valuesEqual = newItem.equals(currentItem);
-                }
+    @NonNull
+    private Observable<ContentProviderOperation> createOperation(@NonNull final U item, @NonNull final Uri uri) {
+        return Observable
+                .fromCallable(() -> {
+                    // We block until this Uri is freed for operations. Failure to lock
+                    // will throw, which we'll catch later.
+                    locker.acquire(uri);
+
+                    final Cursor cursor = contentResolver.query(uri, getProjection(), null, null, null);
+
+                    if (cursor == null || !cursor.moveToFirst()) {
+                        if (cursor != null) {
+                            cursor.close();
+                        }
+
+                        Log.v(TAG, "Create insertion operation for " + uri);
+                        return ContentProviderOperation.newInsert(uri)
+                                .withValues(getContentValuesForItem(item))
+                                .build();
+                    }
+
+                    final U currentItem = read(cursor);
+                    final U newItem = mergedItem(currentItem, item);
+
+                    cursor.close();
+
+                    if (!newItem.equals(currentItem)) {
+                        Log.v(TAG, "Create update operation for " + uri);
+                        return ContentProviderOperation.newUpdate(uri)
+                                .withValues(getContentValuesForItem(newItem))
+                                .build();
+                    }
+
+                    Log.v(TAG, "Data already up to date at " + uri);
+                    return NO_OPERATION;
+                })
+                .onErrorReturn(e -> NO_OPERATION)
+                .doOnNext(operation -> releaseIfNoOp(operation, uri))
+                .filter(this::isValidOperation);
+    }
+
+    private boolean isValidOperation(@NonNull final ContentProviderOperation operation) {
+        return !NO_OPERATION.equals(operation);
+    }
+
+    private void releaseIfNoOp(@NonNull final ContentProviderOperation operation, @NonNull final Uri uri) {
+        if (!isValidOperation(operation)) {
+            try {
+                locker.release(uri);
```

Comment:
Strict code checking rules complain when subclassing Exception, requiring that all user-created exception classes should be unchecked. I guess the idea is that we shouldn't define exceptions inside the app that model errors from outside the application. Changed this to IllegalStateException now, though that's still unchecked.